File: 03_Longest_Substring_Without_Repeating_Characters/Q3_Longest_Substring_Without_Repeating_Characters.js
/**
 * Find the length of the longest substring without repeating characters.
 * @param {string} s - Input string
 * @return {number} Length of the longest substring without repeating characters
 */
var lengthOfLongestSubstring = function(s) {
    const charIndex = {};
    let maxLength = 0;
    let left = 0;
    
    for (let right = 0; right < s.length; right++) {
        // If character is found and is in current window
        if (s[right] in charIndex && charIndex[s[right]] >= left) {
            left = charIndex[s[right]] + 1;
        }
        
        // Update the character index
        charIndex[s[right]] = right;
        
        // Update maxLength
        maxLength = Math.max(maxLength, right - left + 1);
    }
    
    return maxLength;
};

// Main execution
const s = "abcabcbb";
const result = lengthOfLongestSubstring(s);

console.log("Input: " + s);
console.log("Length of longest substring without repeating characters: " + result);

File: 03_Longest_Substring_Without_Repeating_Characters/Q3_Longest_Substring_Without_Repeating_Characters.py
def lengthOfLongestSubstring(s: str) -> int:
    """
    Find the length of the longest substring without repeating characters.
    
    Args:
        s: Input string
        
    Returns:
        Length of the longest substring without repeating characters
    """
    char_index = {}
    max_length = 0
    left = 0
    
    for right in range(len(s)):
        # If character is found and is in current window
        if s[right] in char_index and char_index[s[right]] >= left:
            left = char_index[s[right]] + 1
        
        # Update the character index
        char_index[s[right]] = right
        
        # Update max_length
        max_length = max(max_length, right - left + 1)
    
    return max_length


if __name__ == "__main__":
    s = "abcabcbb"
    result = lengthOfLongestSubstring(s)
    
    print(f"Input: {s}")
    print(f"Length of longest substring without repeating characters: {result}")
